storage-dynamodb/src/main/java/io/delta/storage/DynamoDBLogStore.java

```diff
@@ -16,10 +16,242 @@
+
+    private AmazonDynamoDBClient client = null;
+    private String tableName;
+    private String credentialsProviderName;
+    private String regionName;
+    private String endpoint;
+
+    public DynamoDBLogStore(Configuration hadoopConf) throws IOException {
         super(hadoopConf);
+        tableName = getParam(hadoopConf, "tableName", "delta_log");
+        credentialsProviderName = getParam(
+            hadoopConf,
+            "credentials.provider",
+            "com.amazonaws.auth.DefaultAWSCredentialsProviderChain"
+        );
+        regionName = getParam(hadoopConf, "region", "us-east-1");
+        endpoint = getParam(hadoopConf, "endpoint", null);
+        client = getClient();
+        tryEnsureTableExists(hadoopConf);
+    }
+
+   /*
+    * Write to db in exclusive way.
+    * Method should throw @java.nio.file.FileAlreadyExistsException if path exists in cache.
+    */
+    @Override
+    protected void putExternalEntry(
+        ExternalCommitEntry entry, boolean overwrite
+    ) throws IOException {
+        try {
+            LOG.debug(String.format("putItem %s, overwrite: %s", entry, overwrite));
+            client.putItem(createPutItemRequest(entry, tableName, overwrite));
+        } catch (ConditionalCheckFailedException e) {
+            LOG.debug(e.toString());
+            throw new java.nio.file.FileAlreadyExistsException(
+                entry.absoluteJsonPath().toString()
+            );
+        }
+    }
+
+    @Override
+    protected ExternalCommitEntry getExternalEntry(
+        Path absoluteTablePath, Path absoluteJsonPath
+    ) throws IOException {
+        Map<String, AttributeValue> attributes = new ConcurrentHashMap<>();
+        attributes.put("tablePath", new AttributeValue(absoluteTablePath.toString()));
+        attributes.put("fileName", new AttributeValue(absoluteJsonPath.toString()));
+
+        java.util.Map<String, AttributeValue> item = client.getItem(
+            new GetItemRequest(tableName, attributes)
+               .withConsistentRead(true)
+        ).getItem();
+        return item != null ? itemToDbEntry(item) : null;
+    }
+
+    @Override
+    protected ExternalCommitEntry getLatestExternalEntry(Path tablePath) throws IOException {
+        Map<String, Condition> conditions = new ConcurrentHashMap<>();
+        conditions.put(
+            "tablePath",
+            new Condition()
+                .withComparisonOperator(ComparisonOperator.EQ)
+                .withAttributeValueList(new AttributeValue(tablePath.toString()))
+        );
+
+        try {
+            java.util.Map<String, AttributeValue> item = client.query(
+                new QueryRequest(tableName)
+                    .withConsistentRead(true)
+                    .withScanIndexForward(false)
+                    .withLimit(1)
+                    .withKeyConditions(conditions)
+            ).getItems().get(0);
+            return itemToDbEntry(item);
+        } catch(IndexOutOfBoundsException e) {
```

Comment:
why not check if `getItems()` is empty or not instead?

Reply:
good point